File: analytics/ruby/api_analytics/lib/api_analytics.rb
# frozen_string_literal: true

require 'uri'
require 'net/http'
require 'json'
require 'time'

module Analytics
  class Middleware
    FLUSH_INTERVAL = 60

    def initialize(app, api_key, config = Config.new)
      @app = app
      @api_key = api_key
      @config = config
      @framework = "Middleware"
      @requests = []
      @last_posted = Time.now
      @mutex = Mutex.new
      @flusher_started = false
    end

    def call(env)
      start = Time.now
      status, headers, response = @app.call(env)

      # Collecting analytics must never break the host response, so any failure
      # in a data mapper or the buffer is swallowed here.
      begin
        ensure_flusher

        request_data = {
          hostname: @config.get_hostname.call(env),
          ip_address: get_ip_address(env),
          path: @config.get_path.call(env),
          user_agent: @config.get_user_agent.call(env),
          method: env['REQUEST_METHOD'],
          status: status,
          response_time: ((Time.now - start) * 1000).round,
          user_id: @config.get_user_id.call(env),
          created_at: Time.now.utc.iso8601
        }

        log_request(request_data)
      rescue StandardError
        # Best-effort: drop this request's analytics rather than raise.
      end

      [status, headers, response]
    end

    # flush posts any buffered requests immediately. Exposed so an application
    # can drain the buffer from its own shutdown hook and avoid losing the final
    # batch on a graceful restart.
    def flush
      requests_to_post = nil
      @mutex.synchronize do
        unless @requests.empty?
          requests_to_post = @requests
          @requests = []
          @last_posted = Time.now
        end
      end
      post_requests(requests_to_post) if requests_to_post
    end

    private

    # ensure_flusher lazily starts a background thread that flushes buffered
    # requests every FLUSH_INTERVAL seconds, so a partial batch is not held
    # indefinitely when traffic goes idle. It is started on the first request
    # (rather than at construction) so the thread lives in the worker process
    # after a forking server (e.g. Puma with preload) spawns it.
    def ensure_flusher
      return if @flusher_started

      @mutex.synchronize do
        return if @flusher_started

        @flusher_started = true
      end

      Thread.new do
        loop do
          sleep(FLUSH_INTERVAL)
          begin
            flush
          rescue StandardError
            # Best-effort background flush.
          end
        end
      end
    end

    def get_ip_address(env)
      return nil if @config.privacy_level >= 2

      @config.get_ip_address.call(env)
    end

    def post_requests(requests)
      return if @api_key.to_s.empty?

      payload = {
        api_key: @api_key,
        requests: requests,
        framework: @framework,
        privacy_level: @config.privacy_level
      }

      url = @config.server_url.end_with?('/') ? @config.server_url + 'api/log-request' : @config.server_url + '/api/log-request'
      uri = URI(url)

      Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == 'https', open_timeout: 10, read_timeout: 10) do |http|
        request = Net::HTTP::Post.new(uri, 'Content-Type' => 'application/json')
        request.body = payload.to_json
        http.request(request)
      end
    end

    def log_request(request_data)
      now = Time.now
      requests_to_post = nil

      @mutex.synchronize do
        @requests.push(request_data)
        if (now - @last_posted) > 60.0
          requests_to_post = @requests.dup
          @requests = []
          @last_posted = now
        end
      end

      Thread.new { post_requests(requests_to_post) } if requests_to_post
    end
  end

  Config = Struct.new(:privacy_level, :server_url, :get_path, :get_ip_address, :get_hostname, :get_user_agent, :get_user_id) do
    def initialize(
      privacy_level = 0,
      server_url = 'https://www.apianalytics-server.com',
      get_path = ->(env) { env['REQUEST_PATH'] },
      get_ip_address = lambda { |env|
        env['HTTP_CF_CONNECTING_IP'] ||
          (env['HTTP_X_FORWARDED_FOR'] && env['HTTP_X_FORWARDED_FOR'].split(',').first&.strip) ||
          env['HTTP_X_REAL_IP'] ||
          env['REMOTE_ADDR']
      },
      get_hostname = ->(env) { env['HTTP_HOST'] },
      get_user_agent = ->(env) { env['HTTP_USER_AGENT'] },
      get_user_id = ->(_env) { nil }
    )
      self.privacy_level = privacy_level
      self.server_url = server_url
      self.get_path = get_path
      self.get_ip_address = get_ip_address
      self.get_hostname = get_hostname
      self.get_user_agent = get_user_agent
      self.get_user_id = get_user_id
    end
  end

  class Rails < Middleware
    def initialize(app, api_key, config = Config.new)
      super(app, api_key, config)
      @framework = "Rails"
    end
  end

  class Sinatra < Middleware
    def initialize(app, api_key, config = Config.new)
      super(app, api_key, config)
      @framework = "Sinatra"
    end
  end
end

File: analytics/elixir/api_analytics/lib/api_analytics.ex
defmodule ApiAnalytics do
  @moduledoc """
  API Analytics for Phoenix and Plug applications.

  Add `ApiAnalytics` to your supervision tree (typically in `application.ex`):

      children = [
        {ApiAnalytics, api_key: System.get_env("API_ANALYTICS_KEY")},
        MyAppWeb.Endpoint
      ]

  Then add `ApiAnalytics.Plug` to your endpoint or router pipeline:

      # In endpoint.ex (logs all requests)
      plug ApiAnalytics.Plug

      # Or in router.ex (logs only API pipeline)
      pipeline :api do
        plug :accepts, ["json"]
        plug ApiAnalytics.Plug
      end

  ## Options

    * `:api_key` - Your API Analytics key (required)
    * `:privacy_level` - Controls IP address handling (default: `0`)
      * `0` - IP stored and used for location
      * `1` - IP used for location only, then discarded
      * `2` - IP never sent to the server
    * `:server_url` - Override for self-hosting (default: `"https://www.apianalytics-server.com/"`)
  """

  use GenServer

  @server_url "https://www.apianalytics-server.com/"
  @flush_interval 60_000

  @type t :: %__MODULE__{
          api_key: String.t(),
          privacy_level: non_neg_integer(),
          server_url: String.t(),
          get_path: (Plug.Conn.t() -> String.t()) | nil,
          get_hostname: (Plug.Conn.t() -> String.t()) | nil,
          get_ip_address: (Plug.Conn.t() -> String.t() | nil) | nil,
          get_user_agent: (Plug.Conn.t() -> String.t()) | nil,
          get_user_id: (Plug.Conn.t() -> String.t() | nil) | nil
        }

  defstruct api_key: "",
            privacy_level: 0,
            server_url: @server_url,
            get_path: nil,
            get_hostname: nil,
            get_ip_address: nil,
            get_user_agent: nil,
            get_user_id: nil

  @doc """
  Starts the ApiAnalytics GenServer and links it to the calling process.
  """
  def start_link(opts) do
    {name, opts} = Keyword.pop(opts, :name, __MODULE__)
    GenServer.start_link(__MODULE__, opts, name: name)
  end

  @doc """
  Logs a request. Called automatically by `ApiAnalytics.Plug`.
  """
  def log(request_data, server \\ __MODULE__) do
    GenServer.cast(server, {:log, request_data})
  end

  @impl GenServer
  def init(opts) do
    config = struct(__MODULE__, opts)
    schedule_flush()
    {:ok, %{config: config, requests: []}}
  end

  @impl GenServer
  def handle_cast({:log, request_data}, state) do
    {:noreply, %{state | requests: [request_data | state.requests]}}
  end

  @impl GenServer
  def handle_info(:flush, state) do
    if state.requests != [] do
      flush_requests(state.config, state.requests)
    end

    schedule_flush()
    {:noreply, %{state | requests: []}}
  end

  defp schedule_flush do
    Process.send_after(self(), :flush, @flush_interval)
  end

  defp flush_requests(config, requests) do
    payload =
      Jason.encode!(%{
        api_key: config.api_key,
        requests: Enum.reverse(requests),
        framework: "Phoenix",
        privacy_level: config.privacy_level
      })

    url = String.trim_trailing(config.server_url, "/") <> "/api/log-request"

    :httpc.request(
      :post,
      {String.to_charlist(url), [], ~c"application/json", payload},
      [{:timeout, 10_000}],
      []
    )
  end
end

defmodule ApiAnalytics.Plug do
  @moduledoc """
  Plug middleware that logs each request to `ApiAnalytics`.
  """

  @behaviour Plug

  alias Plug.Conn

  @impl Plug
  def init(opts), do: Keyword.get(opts, :server, ApiAnalytics)

  @impl Plug
  def call(conn, server) do
    start = System.monotonic_time(:millisecond)

    Conn.register_before_send(conn, fn conn ->
      elapsed = System.monotonic_time(:millisecond) - start

      request_data = %{
        hostname: conn.host,
        ip_address: get_ip_address(conn),
        path: conn.request_path,
        user_agent: get_header(conn, "user-agent"),
        method: conn.method,
        status: conn.status,
        response_time: elapsed,
        user_id: nil,
        created_at: DateTime.utc_now() |> DateTime.to_iso8601()
      }

      ApiAnalytics.log(request_data, server)
      conn
    end)
  end

  defp get_ip_address(conn) do
    case Conn.get_req_header(conn, "cf-connecting-ip") do
      [ip | _] ->
        ip

      [] ->
        case Conn.get_req_header(conn, "x-forwarded-for") do
          [forwarded | _] ->
            forwarded |> String.split(",") |> List.first() |> String.trim()

          [] ->
            case Conn.get_req_header(conn, "x-real-ip") do
              [ip | _] -> ip
              [] -> conn.remote_ip |> :inet.ntoa() |> List.to_string()
            end
        end
    end
  end

  defp get_header(conn, header) do
    case Conn.get_req_header(conn, header) do
      [value | _] -> value
      [] -> ""
    end
  end
end
